File: cpp/StyledComputedFactory.cpp
#include "StyledComputedFactory.hpp"
#include "Styled+Equality.hpp"
#include "ShadowTreeUpdateManager.hpp"
#include "Rules.hpp"
#include "Helpers.hpp"
#include "StyleFunction.hpp"
#include "Specificity.hpp"
#include "StyleResolver.hpp"
#include "VariableContext.hpp"

#include <regex>
#include <variant>
#include <vector>
#include <string>
#include <algorithm>
#include <folly/dynamic.h>
#include <NitroModules/AnyMap.hpp>

namespace margelo::nitro::cssnitro {

    using AnyMap = ::margelo::nitro::AnyMap;


    std::shared_ptr<reactnativecss::Computed<Styled *>> makeStyledComputed(
            const std::unordered_map<std::string, std::shared_ptr<reactnativecss::Observable<std::vector<HybridStyleRule>>>> &styleRuleMap,
            const std::string &classNames,
            const std::string &componentId,
            const std::function<void()> &rerender,
            ShadowTreeUpdateManager &shadowUpdates,
            const std::string &variableScope,
            const std::string &containerScope,
            const std::vector<std::string> &validAttributeQueries) {

        // Capture rerender by value (copy) so it persists through fast refresh
        // Capture shadowUpdates by pointer since it's a stable singleton
        auto shadowUpdatesPtr = &shadowUpdates;

        auto computed = reactnativecss::Computed<Styled *>::create(
                [&styleRuleMap, classNames, componentId, rerender, shadowUpdatesPtr, variableScope, containerScope, validAttributeQueries](
                        Styled *const &prev,
                        typename reactnativecss::Effect::GetProxy &get) {
                    Styled *next = new Styled{};
                    std::unordered_map<std::string, AnyValue> mergedStyles;
                    std::unordered_map<std::string, AnyValue> mergedProps;
                    std::unordered_map<std::string, AnyValue> mergedImportantStyles;
                    std::unordered_map<std::string, AnyValue> mergedImportantProps;

                    // Collect all style rules from all classNames
                    std::vector<HybridStyleRule> allStyleRules;

                    std::regex whitespace{"\\s+"};
                    std::sregex_token_iterator tokenIt(classNames.begin(), classNames.end(),
                                                       whitespace, -1);
                    std::sregex_token_iterator end;

                    for (; tokenIt != end; ++tokenIt) {
                        const std::string className = tokenIt->str();
                        if (className.empty()) {
                            continue;
                        }

                        auto styleIt = styleRuleMap.find(className);
                        if (styleIt == styleRuleMap.end() || !styleIt->second) {
                            continue;
                        }

                        const std::vector<HybridStyleRule> &styleRules = get(*styleIt->second);

                        // Add only style rules that pass the test
                        for (const HybridStyleRule &styleRule: styleRules) {
                            if (Rules::testRule(styleRule, get, componentId, containerScope,
                                                validAttributeQueries)) {
                                allStyleRules.push_back(styleRule);
                            }
                        }
                    }

                    // Sort all style rules by specificity (highest specificity first)
                    std::sort(allStyleRules.begin(), allStyleRules.end(),
                              [](const HybridStyleRule &a, const HybridStyleRule &b) {
                                  return Specificity::sort(a.s, b.s);
                              });

                    // Process the inline variables
                    for (const HybridStyleRule &styleRule: allStyleRules) {
                        if (styleRule.v.has_value()) {
                            const auto &inlineVariables = styleRule.v.value();
                            for (const auto &kv: inlineVariables->getMap()) {
                                VariableContext::setVariable(variableScope, kv.first, kv.second);
                            }
                        }

                    }

                    // Process the declarations
                    for (const HybridStyleRule &styleRule: allStyleRules) {
                        if (styleRule.d.has_value()) {
                            const auto &declarations = styleRule.d.value();

                            // Check if this is an important rule (s[0] > 0)
                            const bool isImportant = std::get<0>(styleRule.s) > 0;

                            // Determine which maps to use based on importance
                            auto &targetStyles = isImportant ? mergedImportantStyles : mergedStyles;
                            auto &targetProps = isImportant ? mergedImportantProps : mergedProps;

                            // Process declarations using the helper function
                            StyledComputedFactory::processDeclarations(
                                    declarations, targetStyles, targetProps, get, variableScope);
                        }
                    }

                    // Convert and assign all maps using the helper function
                    if (!mergedStyles.empty()) {
                        next->style = StyledComputedFactory::convertToAnyMap(mergedStyles, true,
                                                                             variableScope, get);
                    }

                    if (!mergedProps.empty()) {
                        next->props = StyledComputedFactory::convertToAnyMap(mergedProps, false,
                                                                             variableScope, get);
                    }

                    if (!mergedImportantStyles.empty()) {
                        next->importantStyle = StyledComputedFactory::convertToAnyMap(
                                mergedImportantStyles, true, variableScope, get);
                    }

                    if (!mergedImportantProps.empty()) {
                        next->importantProps = StyledComputedFactory::convertToAnyMap(
                                mergedImportantProps, false, variableScope, get);
                    }

                    // Only perform these actions if this is a recompute (prev exists)
                    if (prev != nullptr) {
                        // Check if we should rerender before deleting prev
                        if (shouldRerender(*next)) {
                            (void) rerender();
                        }

                        // Notify ShadowTreeUpdateManager of style changes in a batch
                        reactnativecss::Effect::batch([&]() {
                            if (next->style.has_value()) {
                                shadowUpdatesPtr->addUpdates(componentId, next->style.value());
                            }
                            if (next->importantStyle.has_value()) {
                                shadowUpdatesPtr->addUpdates(componentId,
                                                             next->importantStyle.value());
                            }
                        });

                        // Now safe to delete prev
                        delete prev;
                    }

                    return next;
                },
                nullptr);

        return computed;
    }

    bool shouldRerender(const Styled &styled) {
        // Check if props has a value - if it does, we should rerender
        return styled.props.has_value();
    }

    void StyledComputedFactory::processDeclarations(
            const auto &declarations,
            std::unordered_map<std::string, AnyValue> &targetStyles,
            std::unordered_map<std::string, AnyValue> &targetProps,
            reactnativecss::Effect::GetProxy &get,
            const std::string &variableScope) {

        std::visit([&targetStyles, &targetProps, &get, &variableScope](const auto &decl) {
            // decl is a tuple, get the first element (the styles)
            const auto &dStyles = std::get<0>(decl);

            // Check if dStyles is valid before accessing
            if (dStyles) {
                for (const auto &kv: dStyles->getMap()) {
                    // Only set if key doesn't already exist
                    if (targetStyles.count(kv.first) == 0) {
                        // Use StyleResolver to resolve the style value (handles functions, variables, etc.)
                        auto resolvedValue = StyleResolver::resolveStyle(kv.second, variableScope,
                                                                         get);

                        // Skip if resolveStyle returns monostate (unresolved)
                        if (std::holds_alternative<std::monostate>(resolvedValue)) {
                            continue;
                        }

                        targetStyles[kv.first] = resolvedValue;
                    }
                }
            }

            // Check if there's a second element in the tuple (props)
            if constexpr (std::tuple_size<std::decay_t<decltype(decl)>>::value > 1) {
                const auto &dPropsOpt = std::get<1>(decl);

                // dPropsOpt is optional, check if it has a value
                if (dPropsOpt.has_value()) {
                    const auto &dProps = dPropsOpt.value();

                    // Check if dProps is valid before accessing
                    if (dProps) {
                        for (const auto &kv: dProps->getMap()) {
                            // Only set if key doesn't already exist
                            if (targetProps.count(kv.first) == 0) {
                                // Use StyleResolver to resolve the prop value (handles functions, variables, etc.)
                                auto resolvedValue = StyleResolver::resolveStyle(kv.second,
                                                                                 variableScope,
                                                                                 get);

                                // Skip if resolveStyle returns monostate (unresolved)
                                if (std::holds_alternative<std::monostate>(resolvedValue)) {
                                    continue;
                                }

                                targetProps[kv.first] = resolvedValue;
                            }
                        }
                    }
                }
            }
        }, declarations);
    }


    std::shared_ptr<AnyMap> StyledComputedFactory::convertToAnyMap(
            const std::unordered_map<std::string, AnyValue> &mergedMap,
            bool applyStyleMapping,
            const std::string &variableScope,
            reactnativecss::Effect::GetProxy &get) {
        if (applyStyleMapping) {
            // Use StyleResolver's helper function to apply style mapping
            return StyleResolver::applyStyleMapping(mergedMap, variableScope, get);
        } else {
            // For props, just copy all values directly without transform mapping
            auto anyMap = AnyMap::make(mergedMap.size());
            for (const auto &kv: mergedMap) {
                anyMap->setAny(kv.first, kv.second);
            }
            return anyMap;
        }
    }


} // namespace margelo::nitro::cssnitro

File: src/native/js-lib/StyledComputedFactory.ts
/**
 * StyledComputedFactory: Creates computed styles for components
 * Converted from cpp/StyledComputedFactory.cpp
 *
 * Key difference from C++: Always triggers rerender instead of shadow tree updates
 * (no shadow tree access in JS React Native)
 */

import type { AnyMap } from "react-native-nitro-modules";

import type { HybridStyleRule, Styled } from "../../specs/StyleRegistry";
import { Computed } from "./Computed";
import { Effect, type GetProxy } from "./Effect";
import { Observable } from "./Observable";
import { Specificity } from "./Specificity";
import type { StyleEngine } from "./StyleEngineFactory";
import type { AnyValue } from "./types";

/**
 * Create a computed that generates Styled object from class names and style rules
 */
export function makeStyledComputed(
  styleRuleMap: Map<string, Observable<HybridStyleRule[]>>,
  classNames: string,
  componentId: string,
  rerender: () => void,
  variableScope: string,
  containerScope: string,
  validAttributeQueries: string[],
  engine: StyleEngine,
): Computed<Styled | null> {
  const computed = Computed.create<Styled | null>((prev, get) => {
    const next: Styled = {};
    const mergedStyles: Record<string, AnyValue> = {};
    const mergedProps: Record<string, AnyValue> = {};
    const mergedImportantStyles: Record<string, AnyValue> = {};
    const mergedImportantProps: Record<string, AnyValue> = {};

    // Collect all style rules from all classNames
    const allStyleRules: HybridStyleRule[] = [];

    // Split classNames by whitespace
    const classes = classNames.split(/\s+/).filter((c) => c.length > 0);

    for (const className of classes) {
      const styleObs = styleRuleMap.get(className);
      if (!styleObs) {
        continue;
      }

      const styleRules = get(styleObs);

      // Add only style rules that pass the test
      for (const styleRule of styleRules) {
        if (
          engine.rules.testRule(
            styleRule,
            get,
            componentId,
            containerScope,
            validAttributeQueries,
          )
        ) {
          allStyleRules.push(styleRule);
        }
      }
    }

    // Sort all style rules by specificity (highest specificity first)
    allStyleRules.sort((a, b) => (Specificity.sort(a.s, b.s) ? -1 : 1));

    // Process the inline variables
    for (const styleRule of allStyleRules) {
      if (styleRule.v) {
        const inlineVariables = styleRule.v;
        for (const [key, value] of Object.entries(inlineVariables)) {
          engine.variableContext.setVariable(variableScope, key, value);
        }
      }
    }

    // Process the declarations and props
    for (const styleRule of allStyleRules) {
      // Check if this is an important rule (s[0] > 0)
      const isImportant = styleRule.s[0] > 0;

      // Process declarations (styles) from the "d" key
      if (styleRule.d) {
        const targetStyles = isImportant ? mergedImportantStyles : mergedStyles;
        processDeclarations(
          styleRule.d,
          targetStyles,
          get,
          variableScope,
          engine,
        );
      }

      // Process props from the "p" key
      if (styleRule.p) {
        const targetProps = isImportant ? mergedImportantProps : mergedProps;
        processDeclarations(
          styleRule.p,
          targetProps,
          get,
          variableScope,
          engine,
        );
      }
    }

    // Convert and assign all maps
    if (Object.keys(mergedStyles).length > 0) {
      next.style = convertToAnyMap(
        mergedStyles,
        true,
        variableScope,
        get,
        engine,
      );
    }

    if (Object.keys(mergedProps).length > 0) {
      next.props = convertToAnyMap(
        mergedProps,
        false,
        variableScope,
        get,
        engine,
      );
    }

    if (Object.keys(mergedImportantStyles).length > 0) {
      next.importantStyle = convertToAnyMap(
        mergedImportantStyles,
        true,
        variableScope,
        get,
        engine,
      );
    }

    if (Object.keys(mergedImportantProps).length > 0) {
      next.importantProps = convertToAnyMap(
        mergedImportantProps,
        false,
        variableScope,
        get,
        engine,
      );
    }

    // Only perform these actions if this is a recompute (prev exists)
    if (prev !== null) {
      // In JS React Native, we always trigger rerender (no shadow tree)
      // Batch the rerender call to avoid excessive updates
      Effect.batch(() => {
        rerender();
      });
    }

    return next;
  }, null);

  return computed;
}

/**
 * Process declarations (styles or props) and merge into target map
 */
function processDeclarations(
  declarations: AnyMap,
  targetMap: Record<string, AnyValue>,
  get: GetProxy,
  variableScope: string,
  engine: StyleEngine,
): void {
  // Get all key-value pairs from the declarations
  for (const [key, value] of Object.entries(declarations)) {
    // Only set if key doesn't already exist (higher specificity wins)
    if (!(key in targetMap)) {
      // Use StyleResolver to resolve the value (handles functions, variables, etc.)
      const resolvedValue = engine.styleResolver.resolveAnyValue(
        value,
        get,
        variableScope,
      );

      // Skip if resolveStyle returns null/undefined (unresolved)
      if (resolvedValue === null || resolvedValue === undefined) {
        continue;
      }

      targetMap[key] = resolvedValue;
    }
  }
}

/**
 * Convert merged map to AnyMap, applying style mapping if needed
 */
function convertToAnyMap(
  mergedMap: Record<string, AnyValue>,
  applyStyleMapping: boolean,
  variableScope: string,
  get: GetProxy,
  engine: StyleEngine,
): AnyMap {
  if (applyStyleMapping) {
    // Use StyleResolver's helper function to apply style mapping
    return engine.styleResolver.applyStyleMapping(
      mergedMap,
      get,
      variableScope,
      true, // processAnimations = true
    );
  } else {
    // For props, just copy all values directly without transform mapping
    const anyMap: AnyMap = {};
    for (const [key, value] of Object.entries(mergedMap)) {
      if (value !== null && value !== undefined) {
        anyMap[key] = value;
      }
    }
    return anyMap;
  }
}
